add cep for definition of new keys & values in build format

wolfv opened the pull request.

wolfv commented on cep-20.2.md:
```diff
@@ -0,0 +1,651 @@
+conda-build has very hard to understand behavior with multiple outputs. We propose some drastic simplifications.
+Each output in the new format has the same keys as a "top-level" recipe.
+
+Values from the top-level `build` and `about` section are merged into the outputs. The `version` from the top-level
```
Yes, when you have outputs, no outer requirements section is allowed.

beckermr replied:
ohhhhhh I read that as the build env, not the build section. Sorry!

beckermr commented on cep-20.2.md:
```diff
@@ -0,0 +1,651 @@
+        - ${{ compiler('cxx') }}
+        - cmake
+        - ninja
+        - ${{ python if win }}
```
Is python an int value here or a string? Maybe we should require it to be a string in this syntax so people do not have to guess?

wolfv replied:
correct, this is a typo! it should be quoted to be a `jinja` string

wolfv replied:
Fixed, thanks!

marcelotrevisani commented on cep-20.2.md:
```diff
@@ -0,0 +1,651 @@
+  license_url: url
+
+  # REMOVED:
+  # prelink_message:
```
this bit here was added for EULA like stuffs that needs confirmation from the user, I believe NVIDIA or/and intel folks use it? or at least they were keen to have it. I would be careful in removing this one :)

beckermr replied:
Yes I 100% agree. We have to keep this.

beckermr replied:
I suppose we could use prelink scripts? @jakirkham can comment for the nvidia stuff.

wolfv replied:
They aren't using it on conda-forge (I checked)

wolfv replied:
I am also keen on dropping the `post-link`, `pre-link`, and `post-unlink` keys in `build`.

beckermr replied:
The keys or the scripts themselves or both? IDK actually what those keys do in a recipe.

wolfv replied:
The keys in the recipe. We would keep the files.

marcelotrevisani replied:
Prelink_message is mainly for eula purposes that the user needs to accept it, it is not related to the prelink and postlink script that are, as far I'm concerned, deprecated

mbargull replied:
> prelink and postlink script that are, as far I'm concerned, deprecated

post-link and pre-unlink should only be deprecated if similar functionality is offered elsewhere.

> The keys in the recipe. We would keep the files.

Personally, I'm not a big fan of the "some file relative to `meta.yaml`/`recipe.yaml` triggers some implicit behavior" thing.
(I'd even go as far as removing the implicit invocation of `build.sh`/`bld.bat` -- but that might upset people :shrug:.)

kkraus14 replied:
Was there a resolution to this for packages that want / need to display some kind of EULA message or even want to require some kind of interaction from a user, i.e. accepting a EULA?

NVIDIA legal seems to be appeased currently, but from experience these decisions can change, and there will likely be uses outside of NVIDIA packages as well.

jaimergp replied:
It feels like the "if a file named in this particular way is found then we do X" behaviour will stay, but the keys to change how that file is named will not... If we do so, I'd rather we do it like we treat activation scripts (their _target_ location is the trigger, not the source one); but that doesn't work well for pre-link & post-unlink because the files will be gone.

carterbox commented on cep-20.2.md:
```diff
@@ -0,0 +1,651 @@
+      version: ${{ libmamba_version }}
+
+    build:
+      script:
+        script: ${{ "build_mamba.sh" if unix else "build_mamba.bat" }}
```
```suggestion
      script: ${{ "build_mamba.sh" if unix else "build_mamba.bat" }}
```
Was this a mistake?

Also, on a related note, I've always found it confusing that `script:` can either be a string that is the name of a file which is a script or it can be a list of strings which will be converted into a bash or bat script. What can this new format do to make this less confusing?

wolfv replied:
Yep, that was a mistake, thanks! 

I see two ways of achieving this, either by having another level in the dict, e.g.

```yaml
script:
  file: myfile.sh
  # or
  content: |
    ....
```

Or having an additional key, like `script_file` vs `script` or something like that.

Ok, maybe a third one would be something like: `script: ${{ embed("build.sh") }}` and write a jinja function that pulls in the contents of the file. This could be quite useful for other files as well (e.g. `README`, ...).

beckermr replied:
I have to say that I like the current syntax a lot but I get that it is confusing.

dhirschfeld replied:
This could potentially allow choosing the interpreter to use:
```yaml
script:
  shell: bash
  file: build.sh
```

carterbox replied:
> I have to say that I like the current syntax a lot but I get that it is confusing.

@beckermr Could you elaborate on why you like that `script:` is a keyword with two different meanings?

beckermr replied:
I like not having to remember the different names for the different things.

h-vetinari replied:
Actually my first thought when I saw `script: ${{ "build_mamba.sh" if unix else "build_mamba.bat" }}` was: "can't we have a default that abstracts away the different file extensions?"

Something like[^1]:
```
        script: build_mamba  # will use .sh on unix and .bat on win
```

If someone needs differently named scripts, they could always drop down to selectors, but it would make one more source of noise go away in the vast majority of cases.

[^1]: if we need to disambiguate from `script:` for compatibility (is this relevant in a new format that's already breaking things?), we could name the key `script_auto_ext:` or something like that.

h-vetinari replied:
Actually the logic for that could be (since `build_mamba` / `build_mamba.sh` / etc. are just strings): If `$script + ".sh"` exists (on unix; `.bat` on windows) execute it, if it doesn't exist, assume the extension is already in the script, i.e. execute `$script`.

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,650 @@
+          - libmambapy
+          - libmambapy.bindings
+      - script:
+          - python -c "import libmambapy._version; assert libmambapy._version.__version__ == '{{ libmambapy_version }}'"
```
```suggestion
          - python -c "import libmambapy._version; assert libmambapy._version.__version__ == '${{ libmambapy_version }}'"
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,650 @@
+
+          # for some reason tqdm doesn't have a proper colorama dependency so pip check fails
+          # but that's completely unrelated to mamba
+          - python -c "import mamba._version; assert mamba._version.__version__ == '{{ mamba_version }}'"
```
```suggestion
          - python -c "import mamba._version; assert mamba._version.__version__ == '${{ mamba_version }}'"
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,650 @@
+    preserve_egg_dir: bool (default false)
+
+    # only used as a hack to down-prioritize packages
+    track_features: [string]
```
Is it a good moment to rename this (even if we use `track_features` under the hood)?

wolfv replied:
probably!

beckermr replied:
Do we have an actual proposal for what to call the next thing to replace track features? Maybe that has to come first?

wolfv replied:
I would propose `priority` or `downprioritize` and then an integer (positive or negative)

jaimergp commented on cep-20.2.md:
````diff
@@ -0,0 +1,650 @@
+
+```yaml
+# url pointing to the source tar.gz|zip|tar.bz2|...
+url: url
````
This can be a list of URLs too (mirrors of the same package with the same hash; used by R packages)

wolfv replied:
👍

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,650 @@
+  # extra run dependencies
+  run: [MatchSpec]
+# extra files to add to the package for the test
+# Alternatively, use ${{ RECIPE_DIR }} or ${{ SOURCE_DIR }} and a single list?
```
I like this better, imo.

JeanChristopheMorinPerso replied:
I have mixed feelings. I always found `files` and `source_files` ambiguous. Your proposed format would help. I also like your alternatives, but I feel like it could cause ambiguity. For example, if we allow `${{ RECIPE_DIR }}`, does it mean that we can also do `/absolute/path/to/file`?

jaimergp commented on cep-20.2.md:
````diff
@@ -0,0 +1,650 @@
+
+```yaml
+# list of imports to try
+imports: [string]
````
As we move to an language agnostic world, should this be renamed to `python_imports` for clarity?

wolfv replied:
Or we do

```yaml
python:
  imports:
    - mypkg
    - mypkg.foo
```

With the potential to add more keys, e.g. `pytest` or others?

jaimergp replied:
That could work, yes. I think `pytest` would be better handled as a command anyway, but maybe there are other Python specific checks to be made.

kenodegard replied:
I'm thinking renaming `imports` to `python` is sufficient (i.e., language keys are imports):

```
test:
  python:
    - mypkg
    - mypkg.foo
    - mypkg.foo:function
  rust:
    - mypkg
    - mypkg::foo
    - mypkg::foo::function
```

Anything more complex is probably better off in a script?

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,650 @@
+        - ${{ compiler('cxx') }}
+        - cmake
+        - ninja
+        - ${{ "python" if win }}
```
What's the returned value happens when there's no else clause and the if clause is not true? `null`?

wolfv replied:
Yes, a `Yaml::Null` value (that we filter out in these `ConditionalList`).

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,650 @@
+        - yaml-cpp
+        - cli11
+        - fmt
+        - if: win
```
This could also be `${{ "winreg" if win }}` I assume?

wolfv replied:
Exactly.

wolfv replied:
I think that people might prefer the trailing `if` syntax for one liners, and the `if: ...` syntax for lists in lists. What do you think?

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,650 @@
+      - script:
+          - if: unix
+            then:
+              - test -d ${PREFIX}/include/mamba
+              - test -f ${PREFIX}/include/mamba/version.hpp
+              - test -f ${PREFIX}/lib/cmake/libmamba/libmambaConfig.cmake
+              - test -f ${PREFIX}/lib/cmake/libmamba/libmambaConfigVersion.cmake
+              - test -e ${PREFIX}/lib/libmamba${SHLIB_EXT}
```
Are we considering Jinja functions for this type of tests? e.g. `${{ exists() }}`

wolfv replied:
No, I am planning to write one more CEP that gets rid of these :) To implement something like the boa "exists" tests: https://boa-build.readthedocs.io/en/latest/recipe_spec.html#check-for-file-existence-in-the-final-package

wolfv replied:
@jaimergp I made another CEP for the file existence checks. https://github.com/conda-incubator/ceps/pull/57

wolfv commented:
@baszalmstra did some awesome work at making a pydantic model for the schema presented in this PR. What's awesome is that it allows us to get feedback and inline-help in the VSCode editor (check out this video) with all the red squiggly lines:

https://github.com/conda-incubator/ceps/assets/885054/54d247b7-980c-4854-8ecc-3730da23c36d

The repository for the pydantic file and the JSON schema is here: https://github.com/prefix-dev/recipe-format

If you start a new YAML file in VSCode and add the following line to the top, you should have it, too:

```
# yaml-language-server: $schema=https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json
```

We also changed a couple of definitions. Most notably the `script_env` is now:

```yaml
- script_env:
    # list of environment variables to pass through to the script
    passthrough: [string]
    # dictionary of key/value of environment variables to set in the script
    env: {string: string}
    # list of secrets to pass through from the environment
    secrets: [string]
```

In the about section we renamed a couple of fields:

```text
description is now either 
description:
  file: path # path to read description from, e.g. ${SRC_DIR}/README.md

 -- or --  
description: string

renamed:

home -> homepage
doc_url -> documentation
dev_url -> repository
```

And we split the definition of multi-output recipes into a "complex recipe".
Different rules apply to complex aka multi-output recipes:

- toplevel package field can only have a `version` (that is used for all outputs, if they do not define their own version)
- toplevel build field can only have a `number` that is used for all outputs if they do not define their own build-number
- toplevel test is disallowed
- toplevel about and source are still allowed, although we're not sure if we should keep the toplevel source or if that should be done via one of these cache outputs

carterbox commented:
I agree with @AntoinePrv that a new format should be accompanied by a version metadata so that parsers can explicitly query which version of the format they are parsing.

dhirschfeld commented:
> I agree with @AntoinePrv that a new format should be accompanied by a version metadata

Perhaps named `schema_version` (or similar) as `version` is a bit overloaded.

wolfv commented:
An alternative to having the schema_version in the file would be to encode the schema version in the filename (`recipe.v1.yaml` etc...). But I am fine with having an explicit schema version in the recipe (I think docker-compose does it, too).

jezdez commented:
> An alternative to having the schema_version in the file would be to encode the schema version in the filename (`recipe.v1.yaml` etc...). But I am fine with having an explicit schema version in the recipe (I think docker-compose does it, too).

I'm in favor of having it inside the file rather than the filename carry it

mariusvniekerk commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+folder: path
+# hash of the file
+sha256: hex string
+# absolute or relative path from recipe file
```
Should we support absolute paths at all?  Feels kinda weird.

h-vetinari commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+  # if it is only one element and ends with `.sh` or `.bat`
+  script: string | [string]
+
+  # linux only, list of rpaths
+  rpaths: [path] (defaults to ['lib/'])
```
This currently works on osx (though undocumented), but it would be really nice if we could have the same (or similar) functionality on windows, which [comes up](https://github.com/conda/conda-build/issues/4753) for example in adding SP_DIR to the rpath for some packages.

jaimergp commented on cep-20.2.md:
````diff
@@ -0,0 +1,711 @@
+## Build section
+
+```yaml
+build:
````
I would even say that this section should be renamed. The `build` term is too loaded in the ecosystem already, so I'd rather be more specific about this section tries to configure. Something like:

* `build_settings`
* `build_options`
* `build_parameters`

wolfv replied:
I agree that it is confusing to have "build" in the requirements as well as in the "top-level". I wouldn't be opposed ot change it to `settings` or `options`, maybe even without the `build_` prefix.

kenodegard replied:
What if we combine `build` with the `package` section? I've always found it a little odd how the name and version is separate from the number and string

danpetry replied:
+1 for combining build and package. In general I find it somewhat difficult to remember which keys go where in the package/source/build area of the recipe, and it would be nice to clean this area up a bit.

jaimergp replied:
Any decisions made here?

jaimergp commented on cep-20.2.md:
````diff
@@ -0,0 +1,711 @@
+# yaml-language-server: $schema=https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json
+```
+
+## Context section
````
The `context` object in conda and mamba might introduce confusion for those familiar with the term (arguably a small subset, probably conda contributors), but anyway I wonder if there's any value in being more explicit about this and just call it `variables` or `constants`.

wolfv replied:
I think `variables` could be fine. I am also thinking we could allow "variables" to be passed in on the command line ... github actions style :) But maybe that's for later.

danpetry replied:
`config`? Does this section do one and the same thing as the cbc.yaml, i.e. define jinja variables?

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+  include_recipe: bool (defaults to true)
+
+  # environment variables to either pass through to the script environment or set
+  script_env: [env_vars]
```
I always felt this name was very ambiguous (specially due to the presence of `script` elsewhere). `env_vars` is a better term, or even `environment_variables`.

wolfv replied:
I actually need to update the spec because we have already updated this in the recipe-format repository: https://github.com/prefix-dev/recipe-format/blob/f673bbc5778be4ce0d4e2765c89cf45948eb7caa/model.py#L130-L141

Although I just saw that we kept the `script_env` name. Happy to adjust that to something else...

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+# url pointing to the source tar.gz|zip|tar.bz2|...
+url: url
+# destination folder in work directory
+folder: path
```
Another opportunity for a more explicit rename. I initially thought this was the folder inside the artifact that should be extracted or fetched. Instead I'd rather have `destination` or `target_directory` for clarity.

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+  # build time dependencies, in the build_platform architecture
+  build: [MatchSpec]
+  # dependencies to link against, in the target_platform architecture
+  host: [MatchSpec]
```
Echoing the discussion in Element, there's an opportunity here to rename `host` to `target`, and align it with `target_platform`, which IMO better describes the intent.

It's true that it conflates a bit with the [GCC terminology](https://gcc.gnu.org/onlinedocs/gccint/Configure-Terms.html), but in all honesty, the conda ecosystem is mostly non-compiler packages so I am not sure where the value in keeping this relic of autotools is.

danpetry replied:
Without being familiar with GCC terminology, but being familiar with conda recipes, I find build/target/run to be more descriptive of what's going on, especially when those terms are appended with "environment".

isuruf replied:
Please don't change it to `target`. It needlessly confuses things. `target_platform` variable is wrong IMO. It should have been `host_platform.
We need better docs on `build, host, run`.

h-vetinari replied:
I think `target` is way way way easier for people to catch on to and that accepting some ambiguity in the full Canadian cross case is preferable to having this large hurdle to understanding.

That said, we need good docs in either case. I've picked up a lot of bits and pieces over the years, but I've held off on writing docs because of the inevitable "you're wrong about X".

To illustrate what I mean, the most understandable explanation from my POV for the distinction between build and host is the basic cross-compilation case, where the distinction is easy to get (e.g. we have x64 agents to build, but the final package needs to run on <arch>; the build-time dependencies for that architecture are in `host`, etc etc). That explanation is not 100% accurate, but way more effective for transmitting the essentials. However I've never had the energy to face that discussion.

In short: +1 for target

isuruf replied:
I'm -100 for target. There are two different nomenclatures.
1. autotools and conda - build and host
2. cmake etc. - host and target

Now, this suggestion is to introduce a third
3. build and target

Please don't add a third into the mix. Somewhat related meme: https://xkcd.com/927/

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+
+SVN and HG (mercury) source definitions are removed as they are not relevant anymore.
+
+## Requirements section
```
All these environments have `[MatchSpec]` as a value, but we have not standardized what `MatchSpec` is, right?

Right now, we expect the 3-field, space-separated "conda build spec", but there are feature requests to support channel specs there too.

This might be out of scope for this CEP, but for now I'd write a couple lines on the expected syntax:

```
package-name [version [build string]]
```

wolfv replied:
I added a section for this

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+test:
+  - script:
+      - if: unix
+        value:
```
```suggestion
        then:
```

Typo?

carterbox commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+
+## Outputs section
+
+conda-build has very hard to understand behavior with multiple outputs. We propose some drastic simplifications.
+Each output in the new format has the same keys as a "top-level" recipe.
```
I believe that the reason people find multiple-output workflows with conda-build confusing is that it was originally intended to only be used for a "build once, then split" workflow. In that workflow, the outputs [only do one of two things](https://docs.conda.io/projects/conda-build/en/stable/resources/define-metadata.html#specifying-files-to-include-in-output): installation to / removal from the $PREFIX via script, or splitting of files already in the $PREFIX via a glob list. However, people have used the `outputs/script` key for scripts that do MORE than just moving files, but also doing additional build steps. i.e. They try to use the top-level recipe as a cached build which is continued and reused multiple times.

My understanding is that this proposal moves the recipe spec away from "build once, then split" towards a graph of multiple build caches and multiple outputs. The only use case for multi-cache multi-output I know of is a mixed C/Python package where a C library is wrapped in Python extensions which must be recompiled for each Python version. By moving to a recipe specification that explicitly has the concept of cached builds, this makes this type of combined multi-output recipe more efficient because the C library is reused? However, for conda-forge at least, feedstocks run each Python build in parallel, so enabling this type of caching will not be immediately beneficial. In practice, I actually encourage maintainers to use separate recipes in the case described above because it avoids multi-output recipes and migrations of the C library don't trigger unnecessary rebuilds of the python package.

I think the proposed "multi-cache, multi-output" workflow is MORE complicated instead of simpler. An actual simplification of the outputs section would remove the `script` option from outputs and only allow pattern matching (but use REGEX to make it more powerful?). This would enforce a "build once, then split" workflow and avoid the common issues which arise when people try to use the output script for doing additional build steps instead of just moving/selecting files.

Perhaps you could be more specific in this proposal about the problem using the current outputs section, some example use cases where it fails, and the type of workflow you want to enable with the new outputs section.

wolfv replied:
You are right, but you are also missing some of the tricky behavior from conda. Namely that requirements from the top level are mixed into the requirements of the output (that has the same name as the top-level recipe).

Splitting packages is also pretty difficult.
I think people are relatively used to re-use caches e.g. from Github actions. I hope we have an implementation of this soon in rattler-build so you can try it out.

Rebuilding the (existing) C library (as you describe) can hopefully also be avoided if we do some more clever things regarding hashes and comparing if builds already exist.

carterbox replied:
On further reflection, I think that these proposed changes to the outputs section are more than just new keys and values. They are actually implying a fundamental change of the documented (canonical?) conda-build workflow (as opposed to just moving or change the names of exisitng keys) and should be their own CEP.

P.S. I think the flatpak build system is designed to have cached modular components with arbitrary dependency graphs, so maybe we can steal some design or implementation ideas from them?

wolfv replied:
I am not opposed to split out the outputs work into another CEP and leave it undefined in this CEP.
For now we'd still implement as laid out here, but open to changes.

msarahan replied:
> mixed C/Python package where a C library is wrapped in Python extensions which must be recompiled for each Python version

This was one of the main workflows intended. Splitting files up was definitely a major use case (for GCC packages back at this feature's inception), but making wrapper builds in a matrix was equally important (for xgboost back at this feature's inception).

Not that history really matters here. What matters is how hard it is to use, and whether it is saving anyone any time.

wolfv commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+  script_env: [env_vars]
+
+  # A run export adds a dependency to the run requirements of a package if listed in build or host dependencies
+  run_exports: [MatchSpec] OR {strong: [MatchSpec], weak: [MatchSpec], strong_constrains: [MatchSpec], weak_constrains: [MatchSpec], noarch: [MatchSpec]}
```
It's a run export that get's attached to a `noarch` package. This is currently already in conda-build.

If you build a `noarch: python` package and depend on `python` in your host section, the `python` dependency will be added to your run dependencies (as opposed to `python X.X` which happens in the "arch" builds).

E.g. python 3.10 run exports are:

```
weak: python 3.8.*
noarch: python
```

wolfv commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+  # merge the build and host environments (used in many R packages on Windows)
+  merge_build_host: bool
+
+  # ignore run exports by name
+  ignore_run_exports: list
+
+  # ignore run exports coming from the specified packages
+  ignore_run_exports_from: list
```
Hmm, maybe. The run exports section is something that gets embedded into the final build output, while these are ways to control the "other end" of the run exports (which should _not_ be applied). But maybe there is a way to bundle them, idk.

wolfv commented on cep-20.2.md:
```diff
@@ -0,0 +1,711 @@
+# absolute or relative path from recipe file
+patches: [path]
+# removed keys
+# fn: string
```
We need to bring this one back as it can be useful for test files (the opencv-feedstock uses it to put a video in the right spot). We'll rename it to `file_name` though.

wolfv commented:
I just added the definition for the `ScriptEnv` that we want to extend and make more explicit vs. what's currently in conda. There are three values in `ScriptEnv`:

- passthrough: pass env var through from outside to script environment
- env: a map of string -> string of env vars to explicitly set in the environemnt
- secrets: works like passthrough but does not store the env var in the package, and never prints it to stdout / stderr etc.

For passthrough and secrets we can also warn the user if those are not available at build time.

kenodegard commented on cep-20.2.md:
```diff
@@ -0,0 +1,727 @@
+## Motivation
+
+The conda-build format is currently under-specified. For the new format, we try to list all values
+and types in a single document and document them. This is part of a larger effort (see [](/cep-20.1.md) for the new YAML syntax).
```
```suggestion
and types in a single document and document them. This is part of a larger effort (see [CEP-13](/cep-13.md) for the new YAML syntax).
```

kenodegard commented on cep-20.2.md:
````diff
@@ -0,0 +1,727 @@
+
+## Schema version
+
+The implicit version of the YAML schema for a recipe is an integer 1. 
+To discern between the "old" format and the new format, we utilize the file name.
+The old format is `meta.yaml` and the new format is `recipe.yaml`.
+The version can be explicitly set by adding a `schema_version` key to the recipe.
+
+```yaml
+# optional, since implicitly defaults to 1
+schema_version: 1  # integer
+```
````
I wouldn't force a rename of `meta.yaml` (but I'd welcome the option to calling it `recipe.yaml`). Instead, I'd enforce `schema_version>=1` for the newer format, so no `schema_version` implies `0` and is today's underspecified format.

baszalmstra replied:
The problem with that is that the old format is not valid yaml. So to properly know, one would have to use some of the old parsing logic/magic as well.

wolfv commented:
I just pushed a relatively large update to the `build` section. I categorized most options further. I think it helps giving structure to the large number of options. Would love to hear thoughts of others on this. 

Downside is that it makes the migration a bit more complicated.

JeanChristopheMorinPerso commented on cep-20.2.md:
```diff
@@ -0,0 +1,756 @@
+  license: string (SPDX enforced)
+  # the license files
+  license_file: path | [path] (relative paths are found in source directory _or_ recipe directory)
+  # URL that points to the license
```
Should we define what tools are supposed to do with that URL? Is the URL only stored in the artifact as is or should tools actually download the file at the given URL?

JeanChristopheMorinPerso commented on cep-20.2.md:
````diff
@@ -0,0 +1,756 @@
+
+```yaml
+# The context dictionary defines arbitrary key-value pairs for Jinja interpolation
+# and replaces the {% set ... %} commands commonly used in recipes
````
I'm curious, does it accept any valid YAML values? Like can we store maps and lists?
```
context:
  requirements:
    - python
    - requests
    - etc
```

wolfv replied:
For the moment this should only accept scalars.

JeanChristopheMorinPerso replied:
Should this be documented?

danpetry commented on cep-20.2.md:
```diff
@@ -0,0 +1,788 @@
+
+    # used to prefer this variant less
+    # note: was `track_features`
+    down_prioritize_variant: integer (negative, defaults to 0)
```
Or is this logic to keep backwards compatibility with track_features somehow?

wolfv replied:
Yeah this would render into a track-feature, yes (to make the output compatible with `conda` and `legacy-conda` (pre-libmamba).

jaimergp replied:
I feel like `priority` would be confusing due to how the term is used in "channel priority" and such.

danpetry commented on cep-20.2.md:
````diff
@@ -0,0 +1,788 @@
+  run_exports: [MatchSpec] OR RunExports
+```
+
+#### `RunExports` section
````
Has it been considered to remove `run_exports`? It would be nicer, IMHO, if the information for usage of a host or build dependency was captured somehow, and then it was up to the author of the downstream recipe to correctly specify the run dependencies manually. More explicit, easier to debug dependency conflicts, etc.

wolfv replied:
yes, once the new format is done all future changes will be much easier because we can properly lint the YAML and do automatic upgrades. The initial one is the hardest.

There is no problem changing the recipe spec in the future.

danpetry replied:
Thanks, so basically the main thing we're tackling in this chain of CEPs is to remove the semantic comments/jinja flow-control/other stuff that goes away from pure YAML with a few jinja functions and variable substitution?

wolfv replied:
Yeah I would also expect _a lot_ of opposition. Would love to see a CEP with a better idea than run exports though! I am also open to improve the words (weak, strong, etc.) but haven't come up wtih anything better yet (they would still "render" to the same run_exports.json file as today, just different name in the recipe).

msarahan replied:
@danpetry you may not have experienced the compatibility hell before run_exports. The whole idea is that the people who maintain a library (or the package for it) are the best ones to understand how its compatibility changes. If you remove the idea of run_exports, you are instead imposing that knowledge on every consumer of the package. It requires a LOT more knowledge from a lot more people.

People at NVIDIA have asked if we could be more descriptive about where a particular dependency in the rendered recipe came from. This might help with your confusion without sacrificing the centralization of compatibility knowledge.

For example, some verbose mode could output something when a dependency's run_exports are being taken into account. This could just be log output, or it could somehow annotate a rendered version of the recipe.

danpetry commented on cep-20.2.md:
```diff
@@ -0,0 +1,788 @@
+    # what to do when detecting overlinking
+    overlinking_behavior: str # one of "ignore" or "error" (defaults to "error")
+
+  ignore_run_exports:
```
Would this be better off in the `requirements` section, since it's about dependency specification?

JeanChristopheMorinPerso commented on cep-20.2.md:
```diff
@@ -0,0 +1,809 @@
+weak: [MatchSpec]
+# strong constrains adds a run constraint from build -> run_constraints (was `strong_constrains`)
+strong_constraints: [MatchSpec]
+# weak constrains adds a run constraint from host -> run_constrained (was `weak_constrains`)
```
`run_constraints` instead of `run_constrained`?

JeanChristopheMorinPerso commented on cep-20.2.md:
```diff
@@ -0,0 +1,809 @@
+    - xtl >=0.7,<0.8
+  run:
+    - xtl >=0.7,<0.8
+  run_constrained:
```
```suggestion
  run_constraints:
```

mbargull commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+# and replaces the {% set ... %} commands commonly used in recipes
+context:
+  variable: test
+  # note that we can reference previous values, that means that they are rendered in order
+  other_variable: test_${{ variable }}
```
As already requested in the previous CEP:
If we want recipes to conform to the YAML specification (which I strongly suggest we should!), we cannot do this.

Please refer to https://yaml.org/spec/1.2.2/#3221-mapping-key-order :
> In the [representation](https://yaml.org/spec/1.2.2/#representation-graph) model, [mapping keys](https://yaml.org/spec/1.2.2/#nodes) do not have an order. To [serialize](https://yaml.org/spec/1.2.2/#serializing-the-representation-graph) a [mapping](https://yaml.org/spec/1.2.2/#mapping), it is necessary to impose an ordering on its [keys](https://yaml.org/spec/1.2.2/#nodes). This order is a [serialization detail](https://yaml.org/spec/1.2.2/#serializing-the-representation-graph) and should not be used when [composing](https://yaml.org/spec/1.2.2/#composing-the-representation-graph) the [representation graph](https://yaml.org/spec/1.2.2/#representation-graph) (and hence for the preservation of [application](https://yaml.org/spec/1.2.2/#processes-and-models) data). In every case where [node](https://yaml.org/spec/1.2.2/#nodes) order is significant, a [sequence](https://yaml.org/spec/1.2.2/#sequence) must be used. For example, an ordered [mapping](https://yaml.org/spec/1.2.2/#mapping) can be [represented](https://yaml.org/spec/1.2.2/#representation-graph) as a [sequence](https://yaml.org/spec/1.2.2/#sequence) of [mappings](https://yaml.org/spec/1.2.2/#mapping), where each [mapping](https://yaml.org/spec/1.2.2/#mapping) is a single [key/value pair](https://yaml.org/spec/1.2.2/#mapping). YAML provides convenient [compact notation](https://yaml.org/spec/1.2.2/#example-flow-mapping-adjacent-values) for this case.

To rephrase this for your proposal:
If you require some ordering in these mappings, then you're saying the recipes only conform to the specification of serialized YAML but not the YAML specification overall.
Which is not good since we're then again in a situation were we say "the recipes are YAML... but actually, only if you use certain tools to read/write them" (not as bad as "2 template formats on top one serialization format" as current `meta.yaml`, but related).

Meaning, this either has to be a list/sequence (e.g., `[{variable: test}, {other_variable: 'test_${{ variable }}'}]`) or require the consumer of these files to do topological sorting/cycle detection.

wolfv replied:
I think this is no problem. We can easily create the "graph" from traversing the Jinja AST and finding the undefined variables. 

Most languages do have facilities to parse YAML into a Map that is sorted by insertion order (Rust and Python have it, for sure). So we can strongly encourage implementations to write (and read) this data in a sorted order.

mbargull replied:
There is a difference between a format specification and implementation details.
If I do
```python
recipe = yaml.load("recipe.yaml")
recipe["context"] = dict(sorted(recipe["context"].items()))
yaml.dump(recipe, "recipe.yaml")`
```
then this does *not* change the YAML representation as per the YAML specification.

You **do not conform to theYAML specification** if you require this.

> So we can strongly encourage implementations

Yes, I'd also suggest that an implementation should not change the order (having things stable causes less confusion and possible determinism).
But that does not mean the implementations should *have to* do that.
Otherwise, please don't name these files `.yaml`.

To reiterate: This is a specification bug that was already listed amongst others in https://github.com/conda-incubator/ceps/pull/54#pullrequestreview-1515301216 .

wolfv replied:
Sure. We _can_ topologically sort the context object. So we can get rid of this requirement (but still ask implementations to do it because it will be much more logical for humans).

wolfv replied:
I added a note with more details and the request to implementations to do topological sorting.

mbargull commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+  # the build string. This is usually ommited (can use `${{ hash }}` variable here)
+  string: Option<string>, defaults to a build string made from "package hash & build number"
+
+  # A list of conditions under which to skip the build of this package (they are joined by `or`)
+  skip: [list of expressions]
```
(I wouldn't mind if we turn this into a non-negative version (i.e., declare on what to build for, not what to skip) -- but I'd postpone that to some future discussion.)

wolfv replied:
agree, but needs a lot of discussion. Let's do that in a future one.

mbargull commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+  # A list of conditions under which to skip the build of this package (they are joined by `or`)
+  skip: [list of expressions]
+
+  # wether the package is a noarch package, and if yes, wether it is "generic" or "python"
+  noarch: Option<"generic" | "python">
```
It would be a good opportunity to start to split `noarch: python` into `noarch` and `package-type: python` (not that we have to use the `package-type` naming, but you get the gist).

This separation of concerns should be done upstream in `conda` etc. -- but that needs changing of index data format.
It would be good to prepare recipes for such separation, though.
For now, an package building tool would still output `noarch: python` -- but it could simply switch to a new index format once that is available (so that we don't have to change recipes again at that point).

wolfv replied:
I think this needs more careful thoughts. Lets do this in a future CEP. 

We also would like to support Python3 stable ABI packages - they are not `noarch` but still need special hadnling to move the contents to the correct `site-packages` folder.

wolfv replied:
PS this proposal already got rid of the `noarch: true` option :)

mbargull replied:
> I think this needs more careful thoughts. Lets do this in a future CEP.

I would be fine with postponing this -- but it would also be good if we can handle this before the new recipe format sees wider use.

> We also would like to support Python3 stable ABI packages - they are not `noarch` but still need special hadnling to move the contents to the correct `site-packages` folder.

Yes, that makes sense.
As such it would be good to move this into a package-type-specific section (you already proposed `python:` -- which is a good thing!).

wolfv replied:
I think even later we can easily mechanically update the recipes to what you are envisioning.

mbargull commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+    # skip compiling pyc for some files (was `skip_compile_pyc`)
+    skip_pyc_compilation: [glob]
+
+    # This is only used in the pip feedstock.
+    disable_pip: bool (defaults to false)
+
```
Let's outright remove this.
This is more or less just a redundant thing for `conda-build` to do the same as `add_pip_as_python_dependency: false`.
We should not carry these one-off settings over to the new format.
conda-forge already builds with `add_pip_as_python_dependency: false` generally -- if Anaconda or whoever else builds `pip` as a Conda package doesn't do it, they just should.

mbargull commented on cep-20.2.md:
````diff
@@ -0,0 +1,831 @@
+md5: hex string
+# relative path from recipe file
+patches: [path]
+```
````
This is missing the `no_hoist` option.
(Not saying it is a very good option; but that functionality would be lost.)

Examples for where this is used are the CDT packages in conda-forge, e.g., https://github.com/conda-forge/cdt-builds/blob/66808da23b025f89cb40c9c725cf707f9caa4cfa/cdts/alsa-lib-cos7-x86_64/meta.yaml#L8

We could replace this with something better (better name, better functionality), see
https://github.com/conda/conda-build/issues/3595 ,esp., https://github.com/conda/conda-build/issues/3595#issuecomment-509339146 onward.

(An option to prevent the build tool from extracting an archive (can't remember if one exists) would also be good -- but that can be proposed later.)

wolfv replied:
Hmm, I do not understand what no_hoist does. After reading the issue/Pr I get an idea (keep parent folders of the tar around).

I don't find it documented at all: https://docs.conda.io/projects/conda-build/en/latest/search.html?q=no_hoist

I understand that CDT packages are being phased out?

I did also think about an option to prevent the automatic extraction of a tarball. Maybe we should rather do that (and then people can do whatever they want in the build script)?

like an extra option:

```
extract: true / false
or 
auto_extract: true/false (default to true)
```

wolfv replied:
Or `extract_archive: true / false`

wolfv replied:
How about an "extract_archive" option @beckermr? Then in the build script we could still extract the archive with any options available to the `tar` installed in the build env.

beckermr replied:
That will make more work for someone else when they convert recipes with it. I'm not a huge fan of that.

wolfv replied:
No gaiiin without paiiin. More seriously, some of the reorderings here already mean that we will need a more powerful converter.

I relatively strongly disagree with the `no_hoist` option as I don't understand it from the option alone.

In fact in `rattler-build` we don't do any of the "hoisting" anyways. So I am wondering if this is the wrong idea from the start and the default should always be to "not hoist". 

If someone has a deeply nested tar ball then they will just have to cd into the right directory.

jaimergp replied:
Any decisions here? I like `extract`.

mbargull commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+  host: [PackageSelector]
+  # the section below is copied into the index.json and required at package installation
+  run: [PackageSelector]
+  # constrain optional packages (was `run_constrained`)
+  run_constraints: [PackageSelector]
```
Keep in mind that we then have:
- `constrains` as the actual key that is used in the index data later,
- `run_constrained` as the old key in `meta.yaml`,
- `run_constraints` as the new key in `recipe.yaml`.

maybe it makes sense to use a uniform one, i.e., `constrains`(/`run_constrains`)?

(Personally, I don't care much about the naming. But it is bound to confuse people if we use multiple names.)

wolfv replied:
I'll ask this in tonights meeting.

jezdez replied:
Oh, dang, yeah, let's stay consistent

JeanChristopheMorinPerso replied:
https://thecontentauthority.com/blog/constrain-vs-constraint
>‘Constrain’ means to restrict or limit something or someone. It often implies an external force or influence that is imposing the restriction. For example, “The budget constraints prevented us from hiring more staff.”

>‘Constraint’, on the other hand, refers to a limitation or restriction that is inherent in a situation or system. It can also refer to the rules or regulations that govern a particular activity or process. For example, “The constraints of the project timeline meant that we had to work overtime to meet the deadline.”

Here we really mean constraints right?

wolfv replied:
The example for `constrains` uses `constraints`? :D 

I think one can say that these are "added constraints" for the environment. But I am not a native english speaker which is partially why I want to ask in 4 minutes :)

beckermr replied:
The index name is wrong according to my native ears but matches "depends" (which is also not a noun). "Depends" also does not match its corresponding section in requirements/run which is another fun mismatch.

beckermr replied:
The index is really an internal api and so I don't think should influence the recipe spec which is meant for use by folks not developing conda build.

beckermr replied:
Hence I agree with constraints

mbargull commented on cep-20.2.md:
````diff
@@ -0,0 +1,829 @@
+python 3.9.*
+```
+
+> [!NOTE]
+> `MatchSpec` are defined with many more options in conda. We are sticking to conda-build's definition for the time being.
+> For example, `conda` MatchSpecs allow specifying a channel, a build number (in square brackets) and many additional things.
+> We might homogenize these things later on.
````
I think it's fine/good to use a subset for now.
(E.g., in `meta.yaml` you could actually use direct links to package archives or something, but I don't think this is widely supported and will break with `libmamba` solver etc.)

(Also `MatchSpec` is not without issues and we should carefully consider where we use it and whether we might want to replace this with something else generally.)

mbargull commented on cep-20.2.md:
```diff
@@ -0,0 +1,829 @@
+sometimes created the same hashes for two different outputs. E.g. in the following situation, conda-build would only create a single
+`foofoo` package (instead of two variants that exactly pin two different libfoo packages):
+
+<img width="2065" alt="Screenshot 2023-11-01 at 15 47 49" src="https://github.com/wolfv/ceps/assets/885054/0c3ec263-0d00-45a7-b6fc-84e9b8daffa6">
```
Technical detail: Such assets should be copied over alongside the CEP.

jaimergp replied:
+1 and same for line 513

wolfv commented on cep-20.2.md:
```diff
@@ -0,0 +1,829 @@
+
+  # URL to the homepage (used to be `home`)
+  homepage: url
+  # URL to the repository (used to be `dev_url`)
+  repository: url
```
We based this off of cargo. It can be left out if there is no repository.

wolfv replied:
people can also still use the "homepage" url...

mbargull replied:
In any case, I think we should still retain the 3 home/dev/doc (no matter the naming) since it is common for them to be distinct and/or have overlaps between any combination of them.

jaimergp replied:
I'd stick to development too.

wolfv commented on cep-20.2.md:
```diff
@@ -0,0 +1,829 @@
+
+  # the run exports to ignore when calculating the requirements
+  ignore_run_exports:
+    # ignore run exports by name (e.g. `libgcc-ng`)
+    by_name: [string]
+    # ignore run exports that come from the specified packages
+    from_package: [string]
```
Why would you think it refers to a MatchSpec?
It refers to a single package _name_ (string). Same for both of these options.

wolfv replied:
I fail to see the use case for a matchspec here.

mbargull replied:
Oh, I'm not suggesting to use `MatchSpec`.
I just wanted to note that `conda-build` accepts more than just the name in those fields.
Looking at the code at
- https://github.com/conda/conda-build/blob/3.28.0/conda_build/render.py#L418
- https://github.com/conda/conda-build/blob/3.28.0/conda_build/render.py#L238
it seems to (effectively with some manual string handling) mostly concern itself with the names too.
We should just look out for such oddities so that any incompatibility between the old and new recipe formats are deliberate.

----
> Why would you think it refers to a MatchSpec?
> It refers to a single package _name_ (string). Same for both of these options.

I'd hope it refers to just the names :).
Since there is a naming inconsistency (`_name`/`_package`) is it not unreasonable to wonder if there is a behavioral difference too, is all I'm saying.

wolfv commented on cep-20.2.md:
````diff
@@ -0,0 +1,829 @@
+the feedstock-name in the conda-forge case).
+
+```yaml
+# note: instead of `package`, it says `recipe` here
+recipe:
+  name: string # mostly ignored, could be used for feedstock name
+  version: string # merged into each output if not overwritten by output
````
We also debated this. 

One alternative: keep the top-level `package` as is and rename the `package` in the outputs to `output: {name: ..., version: ...}`.

I think repeating the version for each output is somewhat annoying, unfortunately. But maybe I am overestimating how annoying it would be.

msarahan replied:
I think that is pretty annoying. I'd be inclined to remove this, and have the version here come from the top-level package/version key. That would also be the default for any output that didn't explicitly override it. But maybe I'm just set in my ways from the current way of doing things.

jaimergp replied:
+1

wolfv commented:
I just renamed `shared_libraries` to `dynamic_linking` (a suggestion by @mbargull). I think that captures better that also other shared object (such as binaries) are concerned.

However, the other section was called `link_options` and I renamed that to `install_options` which is better differentiated.

Now remaining question is wether we should rename:

```
post_link_script -> post_install_script
pre_unlink_scipt -> pre_uninstall_script
pre_link_message -> pre_install_message
```

wdyt @mbargull ?

wolfv commented:
Two more things that came up:

- remove the `package` key altogether and move `name` and `version` into the outer dictionary. An additional benefit could be that we can inject those two variables "into the context" and make them avaiable as variables which we've been thinking about to remove some duplication. Still not clear how that would interact with outputs, though.
- I just thought that we should add an option into the `install_options` to add environment variables into the activated environment. Or maybe we should add another section for activation scripts and activation env vars. There is a JSON format for env vars that is not widely used and it would be great to migrate most activation scripts to that.

mbargull commented:
> However, the other section was called `link_options` and I renamed that to `install_options` which is better differentiated.
> 
> Now remaining question is wether we should rename:
> 
> ```
> post_link_script -> post_install_script
> pre_unlink_scipt -> pre_uninstall_script
> pre_link_message -> pre_install_message
> ```
> 
> wdyt @mbargull ?

Ah, sorry, missed the notification.
There is some certain nomenclature in `conda` about "install"/"link" -- I'm not too certain how consistently it's used and if I remember correctly, but I think it was "install" meaning the whole installation process and "link" meaning only the file linking things.
So, this would mean that we'd have some diverging nomenclature here.

... That being said, I actually wouldn't mind the name change since I think the "link" naming is probably too technical.
(Other package managers also just call it post-install or something with similar semantics.)

wolfv commented:
@mbargull I removed that section entirely because it is not clear why we have a `pre/post-link` options but no options for activation env vars or scripts. 

The `pre-post link` stuff can also be achieved by dropping the files in the right locations manually during setup. 

We'll come up with another proposal reintroducing these + proper options for activation scripts (I especially want to make it _much_ simpler to set "simple" activation variables using the `json` activation variable thing).

mbargull commented:
> We'll come up with another proposal reintroducing these + proper options for activation scripts (I especially want to make it _much_ simpler to set "simple" activation variables using the `json` activation variable thing).

Sounds good to me!
I think it makes much sense to have these iterations on the format so we end up with a more coherent and user-friendly format.
Could you add a comment on the removed options saying that replacements for them will be re-introduced later?

msarahan commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+
+  # settings concerning only Python
+  python:
+    # PythonEntryPoint: `bsdiff4 = bsdiff4.cli:main_bsdiff4`
```
This was a tiny bit unclear to me whether this parameter accepted either a string or a list of strings. I'm not sure if this addition makes things any clearer, so feel free to reject or modify as you see fit.

```suggestion
    # List of strings, where each string follows this format:
    # PythonEntryPoint: `bsdiff4 = bsdiff4.cli:main_bsdiff4`
```

wolfv commented:
@conda-incubator/steering

This vote falls under the **"Enhancement Proposal Approval"** policy of the [conda governance policy](https://github.com/conda-incubator/governance), please vote and/or comment on this proposal at your earliest convenience.

It needs **60%** of the Steering Council to vote `yes` to pass. To vote, please leave `yes`, `no` or `abstain` as comments below.

If you have questions concerning the proposal, you may also leave a comment or code review. **This vote will end on 2024-01-18, End of Day, [Anywhere on Earth](https://time.is/Anywhere_on_Earth) (AoE).**

Note: This is an extended voting period to accommodate the start of the year and the complexity of the proposal.

baszalmstra commented:
yes

wolfv commented:
yes

jezdez commented:
yes

mariusvniekerk commented:
yes

future machine readable versions of the spec should also clearly state what is optional and what isn't.  Presently the general state is basically everything is optional but not really stated as such.

msarahan commented:
Yes. Nothing blocking. Just some comments.

dhirschfeld reviewed (approved):
Not sure why I've got special status as a listed reviewer, but FWIW :shipit:

xhochy commented:
Yes

ocefpaf commented:
Yes

wolfv commented:
@CJ-Wright @goanpeca @chenghlee @marcelotrevisani @jakirkham @mbargull @kkraus14 @jaimergp this is a reminder to please vote with YES / NO / ABSTAIN on this CEP :)

mbargull commented:
> Could you add a comment on the removed options saying that replacements for them will be re-introduced later?

ping on this

mbargull reviewed (approved):
Yes

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+
+The variables can reference other variables from the context section.
+
+> ![NOTE]
```
```suggestion
> [!NOTE]
```

as per https://github.com/orgs/community/discussions/16925

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+## Context section
+
+The context section is a dictionary with key-value pairs that can be used for string interpolation.
+The right-hand side of the key-value pair is a scalar (bool, number or string).
```
```suggestion
The value must be a scalar (bool, number or string).
```

jaimergp replied:
Can it be `null` too?

jaimergp commented on cep-20.2.md:
````diff
@@ -0,0 +1,832 @@
+```yaml
+# The context dictionary defines arbitrary key-value pairs for Jinja interpolation
+# and replaces the {% set ... %} commands commonly used in recipes
+context:
````
Can conditionals be used here? I can imagine folks wanting to do something like they used to do in meta.yaml with build_numbers:

```yaml
context:
  build_number: 0
  final_build_number: ${{ build_number + 0 if mpi else 0 }}
```

This also makes me wonder whether expressions are allowed in the conditionals or only scalars with no operators or methods.

If not (that's fine), we should probably document it.

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+
+
+
+## Package section
```
Is this section required? If it is, we should write that down.

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  # the build number
+  number: integer  # defaults to 0
+
+  # the build string. This is usually ommited (can use `${{ hash }}` variable here)
```
```suggestion
  # the build string. This is usually ommitted (can use `${{ hash }}` variable here)
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  string: string  # defaults to a build string made from "package hash & build number"
+
+  # A list of Jinja conditions under which to skip the build of this package (they are joined by `or`)
+  # What is valid in a an `if:` condition is valid
```
```suggestion
  # What is valid in an `if:` condition is valid
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  # What is valid in a an `if:` condition is valid
+  skip: [list of expressions]
+
+  # wether the package is a noarch package, and if yes, wether it is "generic" or "python"
```
```suggestion
  # whether the package is a noarch package, and if yes, whether it is "generic" or "python"
```

This can also be `null` right?

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  noarch: OneOf<"generic" | "python">
+
+  # script can be a single string or a list of strings
+  # if script is a single string and ends with `.sh` or `.bat`, then we interpret it as a file
```
```suggestion
  # if script is a single-line string and ends with `.sh` or `.bat`, then we interpret it as a file
```

I guess? :D

jaimergp replied:
Wait, we also have a `script` section below. Isn't this redundant? I'd rather just keep the section and forget about this key in `build`.

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  # if script is a single string and ends with `.sh` or `.bat`, then we interpret it as a file
+  script: string | [string] | Script
+
+  # would love to get rid of this:
```
```suggestion
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+
+  variant:
+    # Keys to forcibly use for the variant computation (even if they are not in the dependencies)
+    use_keys: [string]
```
```suggestion
    use_keys: string | [string]
```

As per line 118, but maybe we do want to only require lists.

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  # settings concerning only Python
+  python:
+    # List of strings, where each string follows this format:
+    # PythonEntryPoint: `bsdiff4 = bsdiff4.cli:main_bsdiff4`
```
I never understood why this is not a map, but I guess the `:` in the syntax makes it confusing for the YAML parser (it would always need to be quoted?).

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+    binary_relocation: bool (defaults to true) | [glob]
+
+    # Allow linking against libraries that are not in the run requirements
+    # (was `missing_dso_whitelist`)
```
<3

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+    # (was `runpath_whitelist`)
+    rpath_allowlist: [glob]
+
+    # what to do when detecting overdepending
+    overdepending_behavior: OneOf<"ignore" | "error"> # (defaults to "error")
+
+    # what to do when detecting overlinking
+    overlinking_behavior: OneOf<"ignore" | "error"> # (defaults to "error")
```
Will `ignore` also print a warning or literally omit the report? Or the calculation? Maybe we need three options:

- skip
- warn
- error

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  # the script environment. You can use Jinja to pass through environment variables
+  # with the `env` key (`${{ env.get("MYVAR") }}`).
+  env: {string: string}
+  # secrets that are set as env variables but never shown in the logs or the environment
```
I'm assuming we only pass the environment variable _name_ here, right?

jaimergp commented on cep-20.2.md:
````diff
@@ -0,0 +1,832 @@
+```yaml
+# file, absolute or relative to recipe file
+path: path
+# if there is a gitignore, adhere to it (or not)
````
Can you elaborate here what this implies? That those files will _not_ be copied over?

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+
+#### Python test element
+
+The python test element renders a `test_import.py` file that contains the imports to test.
```
As an aside note, a naive script with:

```python
import a
import b
import c
```

is not the best because it'll fail early. I'd prefer if the implementations do something like:

```python
errors = []

try:
  import a
except Exception as exc:
  errors.append(("a", exc, traceback))
...

# report all errors
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+#### Python test element
+
+The python test element renders a `test_import.py` file that contains the imports to test.
+It also automatically runs the `pip check` command to check for missing dependencies.
```
This means that `pip` must be always injected in the test environment? Or is it running from the tool (rattler-build) environment? This might cause hard-to-debug conflicts in the test env if the user's package is not compatible with part of the `pip` dependency tree (arguably a very small fraction, but I thought I should bring it up).

chenghlee replied:
I'm also not keen on automatically running `pip check`, especially since optional features and dependencies are a thing. Need to dig out the specific package name, but we have run into issues where `pip check` failed because we (as packagers) explicitly decided to turn off a feature but had no way of informing pip of that change.

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  version: string # merged into each output if not overwritten by output
+
+outputs:
+  - package:
+      name: string
+      version: string (defaults to top-level version)
```
```suggestion
  - name: string
    # version: always points to top-level version, cannot be changed.
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+    requirements:
+      # same definition as top level
+
+    # same definitions as on top level, by default merged from outer recipe
```
I'd define what merge means here:

* The output value overwrites the top-level value
* The two values are merged key-wise
* The two values are merged value-wise

chenghlee replied:
+1

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+
+Before the build, the outputs are topologically sorted by their dependencies. Each output acts as an independent recipe.
+
+> **Note**
```
```suggestion
> [!NOTE]
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+    - ${{ compiler('cxx') }}
+    - cmake
+    - if: unix
+      value: make
```
```suggestion
      then: make
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+          - test -f ${PREFIX}/share/cmake/xtensor/xtensorConfig.cmake
+          - test -f ${PREFIX}/share/cmake/xtensor/xtensorConfigVersion.cmake
+      - if: win
+        value:
```
```suggestion
        value:
```
```suggestion
        then:
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  libmamba_version: "1.4.2"
+  libmambapy_version: "1.4.2"
+  # can also reference previous variables here
+  mamba_version: ${{ libmamba_version }}
```
```suggestion
  mamba_version: "${{ libmamba_version }}"
```

Does the Jinja-like replacement introduce quotes too?

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  number: 0
+
+outputs:
+  - package:
+      name: libmamba
+      version: ${{ libmamba_version }}
```
```suggestion
  - name: libmamba
    # version: always the same as top-level
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+              - cat $PREFIX/include/mamba/version.hpp | grep "LIBMAMBA_VERSION_MINOR ${{ libmamba_version_split[1] }}"
+              - cat $PREFIX/include/mamba/version.hpp | grep "LIBMAMBA_VERSION_PATCH ${{ libmamba_version_split[2] }}"
+
+  - package:
+      name: libmambapy
+      version: ${{ libmambapy_version }}
```
```suggestion
  - name: libmambapy
    # version: always the same as top-level
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+      - script:
+          - python -c "import libmambapy._version; assert libmambapy._version.__version__ == '${{ libmambapy_version }}'"
+
+  - package:
+      name: mamba
+      version: ${{ mamba_version }}
```
```suggestion
  - name: mamba
    # version: always the same as top-level
```

jaimergp commented on cep-20.2.md:
```diff
@@ -0,0 +1,832 @@
+  license_file: LICENSE
+  summary: A fast drop-in alternative to conda, using libsolv for dependency resolution
+  description: Just a package manager
+  repository: https://github.com/mamba-org/mamba
```
```suggestion
  development: https://github.com/mamba-org/mamba
```

jaimergp commented on cep-20.2.md:
````diff
@@ -0,0 +1,832 @@
+  # pre-link: path
+```
+
+### Script section
````
Is this a top-level section or the "value type" for the `script` key in `build`?

chenghlee replied:
Seems to be `build/script` based on the Markdown header depth, but yeah, I think that should be clarified.

jaimergp reviewed (commented):
I have some comments before casting my vote (which will be yes).

* It feels like the current writeup needs some proof-reading. I found some typos and out of date examples that should be sorted out before accepting the CEP.
* Some ongoing conversations have not been resolved with a explicit decision. It's fine if those are not tackled in the present CEP, but at least a comment should be made mentioning that they will (or not) be worked on in the future.
* The top-level `script` vs `build/script` explanation is not clear right now.
* A couple sections could use some more details.
* Some cosmetic changes.

I also have some minor comments that are a matter of taste and uncontroversial (e.g. `path` vs `file`). Feel free to ignore those.

goanpeca commented:
Yes 👍🏼

wolfv commented:
@cj-wright @chenghlee @marcelotrevisani @kkraus14

if you could please vote (even if you vote with abstain), that would be great!!

wolfv commented:
Forgot @jakirkham

marcelotrevisani commented:
Yes

chenghlee commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+
+  # settings concerning the prefix detection in files
+  prefix_detection:
+    # force the file type of the given files to be TEXT or BINARY
+    # for prefix replacement
```
For clarity, we should define what TEXT and BINARY mean in this context.

chenghlee commented on cep-20.2.md:
````diff
@@ -0,0 +1,831 @@
+```yaml
+script:
+  # the interpreter to use for the script
+  interpreter: string  # defaults to bash on UNIX and cmd.exe on Windows
````
Is the intention here to force conda build (or equivalent) to explicitly invoke `bash`, or are we going to assume that `/bin/sh` or `/usr/bin/env sh` is actually bash? (Moving forward, I'm not sure it's safe to assume `sh` and `bash` are equivalent.)

chenghlee commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+  # secrets that are set as env variables but never shown in the logs or the environment
+  # The variables are taken from the parent environment by name (e.g. `MY_SECRET`)
+  secrets: [string]
+  # The file to use as the script. Automatically adds `bat` or `sh` to the filename
+  # on Windows or UNIX respectively (if no file extension is given).
```
I'm not sure I like the idea of automatically appending a extension, especially on *nix where file extensions historically have had no particular semantics. I'm not sure why conda build should care if `build-my-pkg` is written in any particular language, especially if I've already specified `interpreter`. (E.g., if I've provided `interpreter: "/usr/bin/env python"` and specify `file: foo`, I don't want conda build trying to run `python foo.sh`.)

dhirschfeld replied:
If you want to run a python file you should just specify the extension `foo.py`? I think it's probably ok to not support running files without an extension?

chenghlee commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+  # The file to use as the script. Automatically adds `bat` or `sh` to the filename
+  # on Windows or UNIX respectively (if no file extension is given).
+  file: string  # build.sh or build.bat
+  # A string or list of strings that is the script contents (mutually exclusive with `file`)
```
We should define what should happen if both `file` and `content` appear in the same recipe specification file. (Error out vs. one taking priority)

chenghlee commented on cep-20.2.md:
````diff
@@ -0,0 +1,831 @@
+```yaml
+# URL to the git repository or path to local git repository
+git: url | path
+# the following 3 keys are mutually exclusive (branch, tag, and rev)
+# branch to checkout to
+branch: string
+# tag to checkout to
+tag: string
+# revision to checkout to (hash or ref)
+rev: string
````
We should clarify what happens if more than one is specified (error vs precedence).

chenghlee commented on cep-20.2.md:
```diff
@@ -0,0 +1,831 @@
+
+<img width="2065" alt="Screenshot 2023-11-01 at 15 47 49" src="https://github.com/wolfv/ceps/assets/885054/0c3ec263-0d00-45a7-b6fc-84e9b8daffa6">
+
+## About section
```
I would be clear about whether this section (or some subset of it) is _actually_ required. The metadata here is useful for major players in the ecosystem (Anaconda, conda-forge, bioconda, etc.), but for J. Random Person trying to generate a conda package for their own tiny channel, requiring things like an SPDX license field seems overly burdensome.

chenghlee commented:
Left comments, but other than _maybe) making `schema_version: N` required, none of them would block a "yes" vote from me.

CJ-Wright commented:
Yes
I am making this comment solely in my personal capacity and am not conveying any rights to any intellectual property of any third parties.

wolfv commented:
Hi everyone! I have counted the votes:

- 12 yes
- I am counting Cheng as 1 abstain
- No no votes

That means 13 people have voted. The council consists of 15 people. 

That means that: 86% have voted, and 92% have voted yes! This CEP is accepted! 🎉

chenghlee commented:
@wolfv: you can count me as a "yes". :)
